### DISCUSSION TITLE
How should I go about using YARP as an Application Gateway for Kubernetes Cluster?

### DISCUSSION THREAD

**Discussion by DefinitelyADev** [dotnet/yarp#2265] [Q&A] [answered]
Hello! I'm using YARP as a reverse proxy for fewer than five services on my Kubernetes cluster. My knowledge of Kubernetes is limited, primarily gained through theoretical and hobbyist exploration at home, with some exposure through work.

My configuration is similar to this:

```mermaid
graph TD;
   A[APP GATEWAY]<-->B[APP 1];
    A<-->C[APP 2];
    A<-->D[APP 3];
```

The APP GATEWAY manages multiple applications and serves as a proxy, handling requests and transformations for each application. Each application scales independently, leading to varying pod counts.

One challenge I face is that each application has its own ingress with a local domain (e.g., app1.domain.local), causing me to lose health checks at the pod level. My idea is to dynamically update the cluster configuration with the current pod IPs as apps scale, ensuring health checks work for each pod. I'm looking for input on this approach and whether it's feasible. Could the Kubernetes API help achieve this? Is my idea valid? Is it possible to use YARP as an ingress controller (like the project samples) and manage multiple apps through this ingress?

I'm inspired by Azure Application Gateway and want to create something similar, albeit less advanced. Any insights on how to proceed would be greatly appreciated.

**Comment by samsp-msft** [comment, marked as answer]
Have you looked at the kubernetes ingress for YARP? https://github.com/microsoft/reverse-proxy/blob/main/samples/KubernetesIngress.Sample/README.md

**Reply by DefinitelyADev** [reply]
If I understand the readme correctly, when I configure the tank as I would have configured the yarp settings, it will deploy a yarp reverse proxy with those settings as an ingress?

Again my knowledge of kubernetes is not the best. Can I have one yarp ingress with multiple services behind?

**Reply by samsp-msft** [reply]
Yes, you just replicate the paths that you want exposed and map them to their respective k8s services.

``` yaml
spec:
  rules:
    - http:
        paths:
          - path: /foo
            pathType: Prefix
            backend:
              service:
                name: frontend
                port:
                  number: 80
```

**Reply by DefinitelyADev** [reply]
Will take a look at it. Thanks!
